python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,691 @@
+    and those appearing below that occurrence-threshold are discarded.
+    """
+
+    RunnerClass = SingleVisitListTaskRunner
+    ConfigClass = FindDefectsTaskConfig
+    _DefaultName = "findDefects"
+
+    def __init__(self, *args, **kwargs):
+        pipeBase.CmdLineTask.__init__(self, *args, **kwargs)
+        self.makeSubtask("isrForFlats")
+        self.makeSubtask("isrForDarks")
+        # change these back once you use the INFO level logging to check
+        # that all the right things are being done, and only the right things!
+        # self.isrForFlats.log.setLevel(lsstLog.WARN)
+        # self.isrForDarks.log.setLevel(lsstLog.WARN)
+
+        validateIsrConfig(self.isrForFlats, self.config.isrMandatoryStepsFlats,
+                          self.config.isrForbiddenStepsFlats, self.config.isrDesirableSteps)
+        validateIsrConfig(self.isrForDarks, self.config.isrMandatoryStepsDarks,
+                          self.config.isrForbiddenStepsDarks, self.config.isrDesirableSteps)
+        self.config.validate()
+        self.config.freeze()
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Augment argument parser for the FindDefectsTask."""
+        parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+        parser.add_argument("--visitList", dest="visitList", nargs="*",
+                            help=("List of visits to use. Same for each detector."
+                                  " Uses the normal 0..10:3^234 syntax"))
+        parser.add_id_argument("--id", datasetType="newDefects",
+                               ContainerClass=NonexistentDatasetTaskDataIdContainer,
+                               help="The ccds to use, e.g. --id ccd=0..100")
+        return parser
+
+    @pipeBase.timeMethod
+    def runDataRef(self, dataRef, visitList):
+        """Run the defect finding task.
+
+        Find the defects, as described in the main task docstring, from a
+        dataRef and a list of visit(s).
+
+        Parameters
+        ----------
+        dataRef : `lsst.daf.persistence.ButlerDataRef`
+            dataRef for the detector for the visits to be fit.
+        visitList : `list` [`int`]
+            List of visits to be processed. If config.mode == 'VISITS' then the
+        list of visits is used. If config.mode == 'MASTER' then the length of
+        visitList must be one, and the corresponding master calibrations are
+        used.
+
+        Returns
+        -------
+        result : `lsst.pipe.base.Struct`
+            Result struct with Components:
+
+            - ``defects`` : `lsst.meas.algorithms.Defect`
+              The defects found by the task
+            - exitStatus : `int`
+              The exit code
+
+        """
+
+        detNum = dataRef.dataId[self.config.ccdKey]
+        self.log.info("Calculating defects using %s visits for detector %s" % (visitList, detNum))
+
+        defectLists = {'dark': [], 'flat': []}
+
+        if self.config.mode == 'MASTER':
```

Comment:
I am leaving this as-is because @RobertLuptonTheGood has a ticket pending (off this branch - DM-20940) that will address this, so a) it would be duplicating effort that's already been done, and b) will leave him in a rebase nightmare. Consider there an invisible `TODO: DM-20940` on this that's already in progress 🙂

Reply:
Ok. Each of those two new functions feels like a perfect case for a unittest of exactly that function (instead of the integration tests you have below).

Reply:
I trust Robert will do exactly that 🙂